server/src/c0Completions.ts

```diff
@@ -0,0 +1,118 @@
+import * as vscode from "vscode-languageserver";
+import { fromVscodePosition, toVscodePosition } from "./ast";
+import * as nearley from "nearley";
+import * as exp from "./expression-rules";
+import * as parsed from "./parse/parsedsyntax";
+
+export const enum CompletionContextKind {
+  StructAccess,
+  FunctionCall
+}
+
+export type CompletionResult =
+  | StructAccess
+  | FunctionCall
+  | null;
+
+export interface StructAccess {
+  tag: CompletionContextKind.StructAccess;
+  expr: parsed.Expression;
+}
+
+export interface FunctionCall {
+  tag: CompletionContextKind.FunctionCall;
+  name: string;
+  argumentNumber: number;
+}
+
+function scanExpression(source: string, index: number) {
+  let parenStack = 0;
+
+  let pos;
+  for (pos = index; pos >= 0; pos--) {
+    const c = source[pos];
+
+    if (c === ")") parenStack++;
+    if (c === "(") {
+      if (parenStack === 0) break; else parenStack--;
+    }
+    if (c === "=") break;
+    if (c === ";") break;
+    if (c === ",") break;
+
```

Comment:
Should also break on '{' (first expression in a block), and keep track of square brackets in the paren stack (as we could be typing in an array index).

Also, this will parse stuff like x * foo->bar as x * foo, which is not what we want since -> has high precedence, so should break on non-identifier text (A-Za-z0-9_) unless they are spaces at the front or the parenStack > 0.

Reply:
Good point - getting it exactly right is fairly challenging. We also have to be able to deal with nested things like a->b->c. Now that I think of it, it seems the only thing outside of parenthesis that we need to capture is -> as (*a).b will always use parens.

Reply:
It also gets worse for things like

```c
struct foo[] a = alloc_array(struct foo, 123);

a[123].asdfg
```

you have to keep capturing the array

Reply:
Yeah. The codes gonna get messy, but I don't see a better way to handle this than just all these cases right now.